README.md

```diff
@@ -1,7 +1,214 @@
-#TBA...
+# ApiProvider
 
-## Features and bugs
+### Clean and boilerplate free network requests with only a few lines of code.
 
-Please file feature requests and bugs at the [issue tracker][tracker].
+```dart
+  final request = Request(url: Path("https://someapi/user"), httpMethod: HttpMethod.GET);
 
-[tracker]: https://github.com/tikkrapp/api_provider/issues
+  final response = await apiProvider.request<User>(request: request);
+  print("fetched user :${response.body.name}");
+```
+
+### Features:
+
+- Typesafe interface on `Request` and `Response` objects.
+- Powerful `Interceptors`
+- Testing / Mocking out of the Box!
+
+
+
+### Installing
+
+Add the following snippet into your `pubspec.yaml`. (Release on pub will happen when the api is stable)
+
+```yaml
+api_provider:
+    git:
+      url: git://github.com/tikkrapp/api_provider.git
+```
+
+
+
+
+
+### Serializing Requests and Deserializing Responses
+
+Instead of providing a serializer when calling  `request`, all the serialisers are stored in a `SerializeContainer` .The `ApiProvider` will get the stored  `Serializable`'s and `Deserializable` 's that are needed when you call `request`.
+
+⚠️  Important:  `T`  of `request` is used to get the needed serializer, thats why it should never be  `dynamic`.  ⚠️ 
+
+We highly recomment to disable `implicit-dynamic`  and `implicit-casts`.
```

Comment:
In case this has to do with the important line above then we should mark it as that.